core/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java

```diff
@@ -63,12 +61,26 @@ public SearchPhaseExecutionException(StreamInput in) throws IOException {
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         out.writeOptionalString(phaseName);
-        out.writeVInt(shardFailures == null ? 0 : shardFailures.length);
-        if (shardFailures != null) {
+        out.writeVInt(shardFailures.length);
+        for (ShardSearchFailure failure : shardFailures) {
+            failure.writeTo(out);
+        }
+    }
+
+    private static final Throwable deduplicateCause(Throwable cause, ShardSearchFailure[] shardFailures) {
+        if (shardFailures == null) {
+            throw new IllegalArgumentException("shardSearchFailures must not be null");
+        }
+        // if the cause of this exception is also the cause of one of the shard failures we don't add it
+        // to prevent duplication in stack traces rendered to the REST layer
```

Comment:
if the issue is the rest layer, should we instead:
 - keep the cause as-is
 - at serialization time, add a magic number when serializing shard failures if they are the same as the cause of the top-level exception and deserialize by copying the cause
 - at rendering time (toxcontent) not render stack traces of shard failures if they are the same as the top-level exception?